src/rosdistro/manifest_provider/gitlab.py

```diff
@@ -0,0 +1,156 @@
+#    with the distribution.
+#  * Neither the name of Open Source Robotics Foundation, Inc. nor
+#    the names of its contributors may be used to endorse or promote
+#    products derived from this software without specific prior
+#    written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+import json
+import os
+import re
+from urllib.request import urlopen, Request
+from urllib.error import URLError
+from urllib.parse import quote as urlquote
+
+from catkin_pkg.package import parse_package_string
+
+from rosdistro.source_repository_cache import SourceRepositoryCache
+from rosdistro import logger
+
+GITLAB_PRIVATE_TOKEN = os.getenv('GITLAB_PRIVATE_TOKEN', None)
+
+def _gitlab_urlopen(url):
+    req = Request(url)
+    if GITLAB_PRIVATE_TOKEN:
+        req.add_header('Private-Token', GITLAB_PRIVATE_TOKEN)
+    logger.warn('Performing GitLab API query "%s"' % (url,))
+    return urlopen(req)
+
+
+def _gitlab_api_query(server, path, resource, attrs):
+    url = 'https://%s/api/v4/projects/%s/%s' % (server, urlquote(path, safe=''), resource)
+    if attrs:
+        url += '?' + '&'.join(k + '=' + urlquote(str(v), safe='') for k, v in attrs.items())
+    return _gitlab_urlopen(url)
+
+
+def _gitlab_paged_api_query(server, path, resource, attrs):
+    _attrs = {
+        'per_page': 50,
+        **attrs,
+        'pagination': 'keyset',
+        'page': '1',
+    }
+
+    url = 'https://%s/api/v4/projects/%s/%s' % (server, urlquote(path, safe=''), resource)
+    if _attrs:
+        url += '?' + '&'.join(k + '=' + urlquote(str(v), safe='') for k, v in _attrs.items())
+
+    while True:
+        with _gitlab_urlopen(url) as res:
+            for result in json.loads(res.read().decode('utf-8')):
+                yield result
+
+            # Get the URL to the next page
+            links = res.getheader('Link')
+            if not links:
+                break
+            match = re.match(r'.*<([^>]*)>; rel="next"', links)
+            if not match:
+                break
+            url = match.group(1)
+
+
+def gitlab_manifest_provider(_dist_name, repo, pkg_name):
+    assert repo.version
+    server, path = repo.get_url_parts()
+    if not server.startswith('gitlab.'):
```

Comment:
I think it's a minor concern, but there are GitLab instances which don't have gitlab in the domain, e.g. <https://salsa.debian.org>. I think some kind of override or indicator for a GitLab handling via environment variable or YAML type sigil could work as a future enhancement but it's not worth blocking 2 on.

Reply:
I'm also not totally sold on this. Maybe we take this as-is and add an escape hatch later if needed...?

Reply:
Escape hatch could be just an environment variable like `ROSDISTRO_GITLAB_SERVER` that if set has to be an exact match. That allows anyone with a private instance on a funky domain to set it, and any public ones (like salsa) can just be added here as needed.

Reply:
Alright, I added it as suggested, but I also removed the hacky `server.startswith('gitlab.')` in favor of the pattern already used for GitHub, `server.endswith('gitlab.com')` since we don't need that heuristic anymore.